### PR TITLE
numdiff in python

### PR DESCRIPTION
**Pull request by dellaert**
Thanks to @truher  !

### CODE REVIEW

File: python/gtsam/tests/test_numerical_derivative.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
+"""
+GTSAM Copyright 2010-2019, Georgia Tech Research Corporation,
+Atlanta, Georgia 30332-0415
+All Rights Reserved
+
+See LICENSE for the license information
+
+Unit tests for IMU testing scenarios.
```
Comment (dellaert) [inline python/gtsam/tests/test_numerical_derivative.py]: Fix comment

File: python/gtsam/utils/numerical_derivative.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
+"""
+GTSAM Copyright 2010-2019, Georgia Tech Research Corporation,
+Atlanta, Georgia 30332-0415
+All Rights Reserved
+
+See LICENSE for the license information
+
+TestCase class with GTSAM assert utils.
```
Comment (dellaert) [inline python/gtsam/utils/numerical_derivative.py]: Fix comment

File: python/gtsam/utils/numerical_derivative.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        raise TypeError(f"a {type(a)} b {type(b)}")
+    if isinstance(a, np.ndarray):
+        return b - a
+    if isinstance(a, (float, int)):
+        return np.array([[b - a]])  # type:ignore
+    # there is no common superclass for Y
+    return a.localCoordinates(b)  # type:ignore
+
+
+def retract(a, b):
```
Comment (dellaert) [inline python/gtsam/utils/numerical_derivative.py]: rename b and type

File: python/gtsam/utils/numerical_derivative.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        dx[j] = -delta
+        dy2 = local(hx, h(retract(x, dx)))
+        dx[j] = 0
+        H[:, j] = (dy1 - dy2) * factor
+    return H
+
+
+def numericalDerivative21(
+    h: Callable[[X1, X2], Y], x1: X1, x2: X2, delta=1e-5
+) -> np.array:
```
Comment (dellaert) [inline python/gtsam/utils/numerical_derivative.py]: all should be np.ndarray

File: python/gtsam/utils/numerical_derivative.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    h: Callable[[X1, X2, X3, X4, X5, X6], Y],
+    x1: X1,
+    x2: X2,
+    x3: X3,
+    x4: X4,
+    x5: X5,
+    x6: X6,
+    delta=1e-5,
+) -> np.array:
+    return numericalDerivative11(lambda x: h(x1, x2, x3, x4, x5, x), x6, delta)
```
Comment (dellaert) [inline python/gtsam/utils/numerical_derivative.py]: fix line ending

**Review by andre-michelin — APPROVED** [review]

**Pull request merged**